src/api.rs

```diff
@@ -0,0 +1,90 @@
+//! This file contains the API for the runtime GC. Interaction with the GC is
+//! only legal through these functions. 
+//!
+//! The GC runtime consists of a mutable singleton `Collector` struct: 
```

Comment:
I'm not saying we shouldn't do this, but I couldn't immediately work out how we might at some point generalise this to multiple threads. I wonder if the `Collector` struct should appear immutable but have `UnsafeCell`s inside?

Reply:
Perhaps you're right. I did worry about the current approach as it means any method on the collector with a `&mut self` would need to be placed in an unsafe block.

Using an `UnsafeCell` inside a static singleton would mean that the collector would need to be instantiated locally, (inside a `thread_local!` macro). I _think_ this sounds right to me. There will need to be atomic access to the collector, and this will mean that only the collector thread can mutate it.

The Rust code for this isn't very pretty, it would look a bit like this in places:

```rust
thread_local!(static GC: UnsafeCell<usize> = UnsafeCell::new(5));

fn main() {
    // Print the value
    GC.with(|f| {
        println!("{:?}", unsafe {*f.get()})
    });

    // mutate some GC value,
    // Then print the value
    GC.with(|f| {
        unsafe { *f.get() = 10;}
        println!("{:?}", unsafe {*f.get()})
    });
}
```

But at least we'd get finer grain control over where our unsafe boundaries are.

Reply:
What's `with` do?

Reply:
It grants safe access to the thread local:
> The with method yields a reference to the contained value which cannot be sent across threads or escape the given closure.

https://doc.rust-lang.org/std/thread/struct.LocalKey.html

Reply:
`GC` has type `UnsafeCell<...>` so I can't see a `with` function. Maybe I'm being an idiot!

Reply:
`thread_local!` wraps the value in a `LocalKey`.

> This key uses the fastest possible implementation available to it for the target platform. It is instantiated with the thread_local! macro and the primary method is the with method.

Reply:
Aha, I see! I am an idiot :0